### PR TITLE
[installer] support service type `ClusterIP` for proxy

### LINKED ISSUE #9981: Add Option to Set Proxy Service as ClusterIP Type in KOTS Admin UI
## Is your feature request related to a problem? Please describe

For users who don't use the Proxy service as a LoadBalancer and instead have their own LoadBalancer in place, it would be very helpful to have the ability to set the Proxy service as ClusterIP instead. I believe having to manually set this has caused several failed Gitpod updates for me.

## Describe the behaviour you'd like

Ideally a simple toggle in the admin UI would be excellent.

## Describe alternatives you've considered

Manually editing the service for the moment (`kubectl edit svc proxy -n gitpod`).

### PR DESCRIPTION
**Pull request by nandajavarma**
## Description

Currently, there is no way to do a post-processing in `KOTS` like [we suggest to do for replacing service type of `proxy`](https://github.com/gitpod-io/gitpod/blob/main/install/installer/README.md#post-processing-the-yaml) to `ClusterIP`. This PR introduces a new experimental field to the Config so users can provide a patch via the KOTS UI. The patch file can have the following structure:

```yaml
experimental:
  webapp:
    proxy:
      serviceType: "ClusterIP" # allowed values are "ClusterIP", "LoadBalancer", "NodePort", "ExternalName" consistent with https://pkg.go.dev/k8s.io/api/core/v1#ServiceType
```
> There is slight redundancy in the validation procedure at the moment(in the `validation.go` and `service.go` we check the allowed types). This is because we do not validate the config in KOTS as of now. We will remove this redundancy when we have better validation in place.

## Related Issue(s)

Fixes #9981 

## How to test

Follow the instructions of[ this PR ](https://github.com/gitpod-io/gitpod/pull/10536) to setup a self-hosted installation with this change. Go to the `Config` tab of the KOTS UI. In the bottom you will see an `Enable additional options` section. Enable it, and copy paste the aforementioned `YAML` snippet to a file and upload it. Update the installment, and if you do a `kubectl get svc proxy -n gitpod` from the CLI you will see that it now has `ClusterIP` type.

## Release Notes

```release-note
NONE
```

### CODE REVIEW

File: install/installer/pkg/components/proxy/service.go
Comment (adrienthebo) [inline install/installer/pkg/components/proxy/service.go:24]: I see this conditional doesn't have an else clause; if `ServiceType` is misspelled (eg `cluterip`) will this this code default to `ServiceTypeLoadBalancer`? Is there validation occurring in this pipeline?

**Comment by adrienthebo** [comment]
Copying my comment from Slack;

My philosophy on pull request review is to provide constructive feedback and make notes where I see potential changes/improvements/risks, but to grant approval as quickly so the author can land the work when they're ready and adopt feedback at their discretion. I think this is ready to ship but there _might_ be a case where an additional conditional clause (or documentation) could be useful.

Since there are no blocking issues on this PR, is it following team policy to grant an approval here? And if we grant that approval, is that sufficient to merge the PR?

File: install/installer/pkg/components/proxy/service.go
Comment (nandajavarma) [reply install/installer/pkg/components/proxy/service.go:24]: This is a great point Adrien! I agree that we can add a validation process. As of this implementation, it just defaults to `LoadBalancer` as set in [this line](https://github.com/gitpod-io/gitpod/blob/main/install/installer/pkg/config/v1/experimental/validation.go).

About the validation, it is a great tip! We can leverage the existing [`validate` cmd](https://github.com/gitpod-io/gitpod/blob/main/install/installer/pkg/config/v1/experimental/validation.go)! I will do a follow-up PR for this!

**Review by adrienthebo — APPROVED** [review]
LGTM! 👍

**Comment by corneliusludmann** [comment]
> My philosophy on pull request review is to provide constructive feedback and make notes where I see potential changes/improvements/risks, but to grant approval as quickly so the author can land the work when they're ready and adopt feedback at their discretion. I think this is ready to ship but there _might_ be a case where an additional conditional clause (or documentation) could be useful.

100 % agree with this philosophy!


> Since there are no blocking issues on this PR, is it following team policy to grant an approval here? And if we grant that approval, is that sufficient to merge the PR?

Approving PR is fine! :heavy_check_mark: However, since we automatically merge PRs by a bot once all required approvals are given, you should add the hold label on this PR so that the author has the opportunity to address your comments. You can do this with this command in a comment:

/hold

The author just need to remove the hold label or comment with `/unhold` and the PR will be merged.

(In this case, the PR still needs approval from Team WebApp. That's because this PR has not been merged yet automatically.)

File: install/installer/pkg/config/v1/experimental/experimental.go
Comment (geropl) [inline install/installer/pkg/config/v1/experimental/experimental.go:35]: If this is special to proxy: Why not add it to `ProxyConfig` ([link](https://github.com/gitpod-io/gitpod/pull/10537/files#diff-7480b9126452b70b6c97afd8fbdcf95a8b614a986c02c113aafc5f691fef5630R165)), then?

Mentioning this because we have tried to keep things simple and non-generic in the installer to start with. And only extend to general solutions where it solved multiple use-cases. But I don't think we have another service that needs the same type of variance in the near future.

Comment (geropl) [reply install/installer/pkg/config/v1/experimental/experimental.go:35]: Also happy to sync to unblock here! :slightly_smiling_face:

Comment (nandajavarma) [reply install/installer/pkg/config/v1/experimental/experimental.go]: Yikes, I acted on a whim :see_no_evil: Have updated the setup now. Sorry for the back-and-fro.

File: install/installer/pkg/components/proxy/service.go
Comment (geropl) [inline install/installer/pkg/components/proxy/service.go:31]: Do we need to tie this to the `ServiceType`? :thinking: Really not sure what happens if we set this field, but do not have `LoadBalancer` type.

Comment (nandajavarma) [reply install/installer/pkg/components/proxy/service.go:31]: You mean the bit here right? https://github.com/gitpod-io/gitpod/pull/10537/files#diff-d9a72ab10ad2fef022eeb810eac2018092ca6dec333c9c38c5c5c8c5244bafc0R79
It doesn't misbehave if we leave the `loadbalancerIP` in, I just didn't want to make the service definition confusing. I can roll this back.

Comment (nandajavarma) [reply install/installer/pkg/components/proxy/service.go:31]: :heavy_check_mark:

Comment (nandajavarma) [reply install/installer/pkg/components/proxy/service.go:31]: Scratch my comment, now I get what you mean. https://github.com/gitpod-io/gitpod/pull/10537/files#diff-d9a72ab10ad2fef022eeb810eac2018092ca6dec333c9c38c5c5c8c5244bafc0R79 this saves from setting the `loadbalancerIP` if the `ServiceType` is not loadbalancer. This reduces any potential confusion, although shouldn't effect the service in any ways.

I put  the check logic here just because the `webapp` experimental section was being parsed anyway, wanted to make the bust of the if jump.

Comment (geropl) [reply install/installer/pkg/components/proxy/service.go:31]: >  this saves from setting the loadbalancerIP if the ServiceType is not loadbalancer.

Ah, thanks, missed that! :bow:

File: install/installer/pkg/config/v1/experimental/experimental.go
Comment (geropl) [reply install/installer/pkg/config/v1/experimental/experimental.go]: Awesome, thx for addressing! :pray:

**Review by geropl — APPROVED** [review]
:heavy_check_mark:

**Pull request merged**
